File: lox-zig/src/lox.zig
const Lox = @This();
const Token = @import("token.zig");
const std = @import("std");

had_error: bool = false,

pub fn run(lox: *Lox, src: []const u8, allocator: std.mem.Allocator) !void {
    var tokens = std.ArrayList(Token).init(allocator);
    defer tokens.deinit();
    var scanner = Token.Scanner.init(src);
    while (scanner.next()) |res| {
        switch (res) {
            .ok => |token| try tokens.append(token),
            .err => |scan_err| {
                switch (scan_err) {
                    .unexpected_character => |line| lox.err(line, "Unexpected character."),
                    .unterminated_string => |line| lox.err(line, "Unterminated string."),
                }
            },
        }
    }
    try tokens.append(.{
        .token_type = .eof,
        .lexeme = "\x00",
        .line = scanner.line + 1,
    });
    const stdout_file = std.io.getStdOut().writer();
    var bw = std.io.bufferedWriter(stdout_file);
    const stdout = bw.writer();
    for (tokens.items) |value| {
        try stdout.print("{}\n", .{value});
    }
    try bw.flush();
}

pub fn err(lox: *Lox, line: usize, msg: []const u8) void {
    lox.report(line, "", msg);
}

fn report(lox: *Lox, line: usize, where: []const u8, msg: []const u8) void {
    std.debug.print("[line {d}] Error{s}: {s}\n", .{ line, where, msg });
    lox.had_error = true;
}

File: lox-rust/src/lox.rs
use crate::token::{Scanner, ScannerErr};

pub struct Lox {
    had_error: bool,
}

impl Lox {
    pub fn new() -> Self {
        Self { had_error: false }
    }

    pub fn run(&mut self, src: &str) {
        use ScannerErr::*;
        let mut tokens: Vec<_> = Scanner::new(src)
            .filter_map(|res| {
                match res {
                    Ok(token) => Some(token),
                    Err(err) => {
                        match err {
                            UnexpectedCharacter(line) => self.error(line, "Unexpected character."),
                            BadUtf8(line) => self.error(line, "Bad UTF-8."),
                            UnterminatedString(line) => self.error(line, "Unterminated string."),
                        };
                        None
                    },
                }
            })
            .collect();
        let line = match tokens.last() {
            Some(t) => t.line + 1,
            None => 1,
        };
        tokens.push(crate::token::Token {
            token_type: crate::token::Type::EOF,
            lexeme: "\0",
            line,
        });
        for token in tokens.into_iter() {
            println!("{token}");
        }
    }

    pub fn error(&mut self, line: usize, msg: &str) {
        self.report(line, "", msg);
    }

    fn report(&mut self, line: usize, err_where: &str, msg: &str) {
        eprintln!("[line {line}] Error{err_where}: {msg}");
        self.had_error = true;
    }

    pub fn had_error(&self) -> bool {
        self.had_error
    }

    pub fn reset_error(&mut self) {
        self.had_error = false;
    }
}
